distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java

```diff
@@ -0,0 +1,379 @@
+        if (!isValid())
+            throw new IllegalArgumentException();
+    }
+
+    /**
+     * Constructor of KadDAction given SMSMessage.
+     *
+     * @param buildingMessage the given SMSMessage.
+     * @throws IllegalArgumentException if the parameters are not valid.
+     */
+    public KadAction(@NonNull SMSMessage buildingMessage) {
+        String messageBody = buildingMessage.getData();
+        actionPeer=buildingMessage.getPeer();
+        try{
+        actionType=ActionType.getTypeFromVal(Integer.parseInt(messageBody.substring(ACTION_TYPE_START_INDEX,OPERATION_ID_START_INDEX)));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(e.getMessage() + "\n" + ACTION_CODE_NOT_FOUND_ERROR_MSG);
+        }
+        operationId=removePadding(messageBody.substring(OPERATION_ID_START_INDEX,CURRENT_PART_START_INDEX));
+        currentPart=removePadding(messageBody.substring(CURRENT_PART_START_INDEX,TOTAL_PARTS_START_INDEX));
+        totalParts=removePadding(messageBody.substring(TOTAL_PARTS_START_INDEX,PAYLOAD_TYPE_START_INDEX));
+        try{
+        payloadType=PayloadType.getTypeFromCode(Integer.parseInt(messageBody.substring(PAYLOAD_TYPE_START_INDEX,PAYLOAD_START_INDEX)));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException(e.getMessage() + "\n" + PAYLOAD_TYPE_NOT_FOUND_ERROR_MSG);
+        }
+        payload=messageBody.substring(PAYLOAD_START_INDEX);
+    }
+
+    /**
+     * Check if all the action's parameters are valid
+     *
+     * @return True if the defined action is valid and fits into a Message.
+     */
+    @Override
+    public boolean isValid() {
+        if (this.actionType.equals(ActionType.INVALID))
+            return false;
+        if (this.payloadType.equals(PayloadType.INVALID))
+            return false;
+        if (currentPart <MIN_PARTS || currentPart > totalParts || totalParts>MAX_PARTS )
+            return false;
+        if(operationId<MIN_ID || operationId>MAX_ID)
+            return false;
+        if (!payloadMatchesType(payloadType, payload))
+            return false;
+        try {
+            toMessage();
+        } catch (InvalidMessageException e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Check if payload's content type matches with payloadType passed as parameter.
+     *
+     * @param payloadType The type of the payload.
+     * @param payload The string which contains the value of the payload.
+     * @return True if the payload string contains a value of the correct type,false otherwise.
+     */
+    public static boolean payloadMatchesType(PayloadType payloadType, String payload) {
+        switch (payloadType) {
+            case IGNORED:
+            case BOOLEAN:
+                return true;
+            case NODE_ID:
+                BitSet bitset=BitSetUtils.decodeHexString(payload);
+                return bitset.size()==ID_LENGTH;
```

Comment:
I'm so very sorry I've seen this just now, you can't do this control, because bitsets aren't of fixed length.
Just check if you can build the bitset (no exceptions) and if size <= ID_LENGTH.
This happens because BitSets only save the positions of bits set to 1.
Btw I updated BinarySet some time ago adding a contructor from hexadecimal string

Reply:
I noticed it just now by testing the builder